File: numerical/python/domain.py
import torch
import math


class Domain:


    def __init__(self, width, height, meter_per_cell):
        self.width = width
        self.height = height

        # Create a torch grid for h, S, u and v
        # h: Water layer thickness
        # S: Sediment bed
        # u: shoreward depth-averaged flow velocity
        # v: alongshore depth-averaged flow velocity
        self.h = torch.zeros(self.height, self.width)
        self.S = torch.zeros(self.height, self.width)
        self.u = torch.zeros(self.height, self.width)
        self.v = torch.zeros(self.height, self.width)
        self.B = torch.zeros(self.height, self.width)

        self.Hin = 1e-5
        self.Hc = 1e-3
        self.H0 = 0.02 # Initial water thickness
        self.grav = 9.81
        self.rho = 1000 # Water density
        self.Du = 0.5 # Turbulent Eddy velocity
        self.nb = 0.016 # bed roughness for bare land
        self.nv = 0.2 # bed roughness for vegetated land
        self.k = 1500 # Vegetation carrying capacity
        self.D0 = 1e-7 # Sediment diffusivity in absence of vegetation
        self.pD = 0.99 # fraction by which sediment diffusivity is reduced when vegetation is at carrying capacity
        self.Sin = 5e-9 # Maximum sediment input rate
        self.Qs = 6e-4 # water layer thickness at which sediment input is halved
        self.Es = 2.5e-4 # Sediment erosion rate
        self.pE = 0.9 # Fraction by which sediment erosion is reduced when vegetation is at carrying capacity
        self.r = 3.2e-8 # Intrinsic plant growth rate (=1 per year)
        self.Qq = 0.02 # Water layer thickness at which vegetation growth is halved
        self.EB = 1e-5 # Vegetation erosion rate
        self.DB = 6e-9 # Vegetation diffusivity
        self.morphological_acc_factor = 44712 # Morphological acceleration factor, required for S and B
        self.pEst = 0.002 # Probability of vegetation seedling establishment


        # Domain properties
        self.dx = meter_per_cell
        self.dy = meter_per_cell
        self.xlen = self.width * self.dx
        self.ylen = self.height * self.dy

    #
    # INITIAL CONDITIONS (PRESETS)
    #
    def init_preset_gaussian_wave(self, x_mu, y_mu, x_sig, y_sig, correlation=0):
        x = torch.linspace(0, self.xlen, self.width)
        y = torch.linspace(0, self.ylen, self.height)
        x, y = torch.meshgrid(x, y, indexing='xy')

        A = 1 / (2*math.pi * x_sig * y_sig * (1-correlation**2)**0.5)

        self.h[:, :] = 1 + A * torch.exp(
            - (1/(2*(1-correlation**2))) * (((x-x_mu)/x_sig)**2 - 2*correlation*((x-x_mu)/x_sig)*((y-y_mu)/y_sig) + ((y-y_mu)/y_sig)**2)
        )

    def initialize(self):
        # Sedimentary elevation is zero everywhere
        self.S[:, :] = 0

        # Flow velocities are zero everywhere
        self.u[:, :] = 0
        self.v[:, :] = 0

        # Uniform water thickness H0
        self.h[:, :] = self.H0

        # Vegetation density is zero everywhere, except for some randomly placed tussocks
        self.B[:, :] = 0
        self.B[torch.where(torch.rand(self.height, self.width) < self.pEst)] = self.k

    #
    # MEMORY MANAGEMENT
    #
    def to(self, device):
        self.h = self.h.to(device)
        self.S = self.S.to(device)
        self.u = self.u.to(device)
        self.v = self.v.to(device)
        self.B = self.B.to(device)

    def get_h(self):
        return self.h.cpu().numpy()

    def get_S(self):
        return self.S.cpu().numpy()

    def get_B(self):
        return self.B.cpu().numpy()

    def get_u(self):
        return self.u.cpu().numpy()

    def get_v(self):
        return self.v.cpu().numpy()

File: numerical/rust/saltmarsh/src/domain.rs
use rand::prelude::*;


pub struct Domain {
    // Domain properties
    width: usize,
    height: usize,
    dx: f64,
    dy: f64,

    // Domain fields
    h: Vec<Vec<f64>>,
    s: Vec<Vec<f64>>,
    u: Vec<Vec<f64>>,
    v: Vec<Vec<f64>>,
    b: Vec<Vec<f64>>,

    // Parameters
    h_in: f64,
    h_c: f64,
    h_0: f64,
    grav: f64,
    rho: f64,
    d_u: f64,
    nb: f64,
    nv: f64,
    k: f64,
    d_0: f64,
    p_d: f64,
    s_in: f64,
    q_s: f64,
    e_s: f64,
    p_e: f64,
    r: f64,
    q_q: f64,
    e_b: f64,
    d_b: f64,
    morphological_acc_factor: f64,
    p_est: f64

}


impl Domain {
    pub fn new(width: usize, height: usize, cell_size: f64) -> Domain {
        // Prepare data fields
        let h = vec![vec![0.0; width]; height];
        let s = vec![vec![0.0; width]; height];
        let u = vec![vec![0.0; width]; height];
        let v = vec![vec![0.0; width]; height];
        let b = vec![vec![0.0; width]; height];

        Domain {
            width,
            height,
            dx: cell_size,
            dy: cell_size,
            h,
            s,
            u,
            v,
            b,
            h_in: 1e-5,
            h_c: 1e-3,
            h_0: 0.02,    // Initial water thickness
            grav: 9.81,
            rho: 1000.0,   // Water density
            d_u: 0.5, // Turbulent Eddy velocity
            nb: 0.016,   // bed roughness for bare land
            nv: 0.2, // bed roughness for vegetated land
            k: 1500.0, // Vegetation carrying capacity
            d_0: 1e-7,    // Sediment diffusivity in absence of vegetation
            p_d: 0.99,    // fraction by which sediment diffusivity is reduced when vegetation is at carrying capacity
            s_in: 5e-9,   // Maximum sediment input rate
            q_s: 6e-4,    // water layer thickness at which sediment input is halved
            e_s: 2.5e-4,  // Sediment erosion rate
            p_e: 0.9, // Fraction by which sediment erosion is reduced when vegetation is at carrying capacity
            r: 3.2e-8,   // Intrinsic plant growth rate (=1 per year)
            q_q: 0.02,    // Water layer thickness at which vegetation growth is halved
            e_b: 1e-5,    // Vegetation erosion rate
            d_b: 6e-9,    // Vegetation diffusivity
            morphological_acc_factor: 44712.0, // Morphological acceleration factor, required for S and B
            p_est: 0.002 // Probability of vegetation seedling establishment
        }
    }

    //
    // Initialize domain with random vegetation tussocks
    //
    pub fn initialize(&mut self) {
        for y in 0..self.height {
            for x in 0..self.width {
                // Sedimentary elevation is zero everywhere
                self.s[y][x] = 0.0;

                // Flow velocities are zero everywhere
                self.u[y][x] = 0.0;
                self.v[y][x] = 0.0;

                // Uniform water thickness H0
                self.h[y][x] = self.h_0;

                // Vegetation density is zero everywhere, except for some randomly placed tussocks
                // Fill the array with uniform random numbers


            }
        }
    }




}